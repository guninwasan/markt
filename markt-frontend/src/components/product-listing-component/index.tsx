export * from "./product-listing-component";
