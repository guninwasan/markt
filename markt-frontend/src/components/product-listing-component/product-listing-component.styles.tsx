import styled from "styled-components";

type ProductListingComponentProps = {
  isMobile: boolean | null;
};

const ProductListingContainer = styled.div<ProductListingComponentProps>`
  display: flex;
  flex-direction: ${({ isMobile }) => (isMobile ? "column" : "row")};
  gap: 2rem;
  padding: 2rem;
`;

const ProductImages = styled.div`
  flex: 1;
`;

const ProductDetails = styled.div<ProductListingComponentProps>`
  flex: 2;
  display: flex;
  flex-direction: column;
  gap: 1rem;
`;

const TitleAndPriceContainer = styled.div<ProductListingComponentProps>`
  display: flex;
  flex-direction: ${({ isMobile }) => (isMobile ? "column" : "row")};
  gap: 2rem;
`;

const TitleAndDescription = styled.div`
  flex: 3;
  display: flex;
  flex-direction: column;
  gap: 1rem;
`;

const PriceBox = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 1rem;
`;

const PriceText = styled.div`
  font-weight: bold;
  font-size: 2.5rem;
`;

const SellerInfo = styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;
`;

const SellerAvatar = styled.img`
  width: 50px;
  height: 50px;
  border-radius: 50%;
`;

const ProductSpecsContainer = styled.div`
  margin-top: 2rem;
  border: 1px solid #ddd;
  padding: 1rem;
  border-radius: 8px;
  background-color: #f9f9f9;
`;

const BottomTab = styled.div<{ isVisible: boolean }>`
  position: fixed;
  bottom: 0;
  width: 100%;
  background-color: white;
  border-top: 1px solid #ddd;
  padding: 1rem;
  display: ${({ isVisible }) => (isVisible ? "flex" : "none")};
  justify-content: center;
  align-items: center;
  box-shadow: 0 -2px 5px rgba(0, 0, 0, 0.1);
`;

const ShareInterestButton = styled.button`
  padding: 0.8rem 1.5rem;
  font-size: 1rem;
  font-weight: bold;
  color: white;
  background-color: #007bff; 
  border: none;
  border-radius: 5px;
  cursor: pointer;

  &:hover {
    background-color: #0056b3; 
  }

  &:focus {
    outline: none;
  }
`;

const ModalBackdrop = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1000;
`;

const ModalContent = styled.div`
  background: white;
  padding: 2rem;
  border-radius: 8px;
  width: 90%;
  max-width: 500px;
  text-align: left; 

  p {
    margin-bottom: 1.5rem;
    font-size: 1rem;
    font-weight: normal; 
    font-style: italic;
  }

  button {
    padding: 0.5rem 1rem;
    font-size: 1rem;
    font-weight: bold;
    color: white;
    background-color: #007bff; 
    border: none;
    border-radius: 5px;
    cursor: pointer;

    &:hover {
      background-color: #0056b3;
    }

    &:focus {
      outline: none;
    }
  }
`;


const NoteTextArea = styled.textarea`
  width: 100%;
  height: 100px;
  padding: 0.5rem;
  margin-top: 1rem;
  font-size: 1rem;
  border: 1px solid #ddd;
  border-radius: 5px;
  resize: vertical;
  font-family: inherit; 

  &:focus {
    outline: none;
    border-color: #007bff;
  }
`;


const SpecGrid = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
  gap: 20px;
`;

const SpecCategory = styled.div`
  background: #f9f9f9;
  padding: 15px;
  border-radius: 8px;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  overflow: hidden;
  h3 {
    margin-bottom: 10px;
    font-size: 1.1em;
    color: #333;
    border-bottom: 2px solid #ddd;
    padding-bottom: 5px;
  }
`;

const SpecItem = styled.p`
  margin: 5px 0;
  font-size: 0.95em;
  color: #555;
`;

export {
  ProductListingContainer,
  ProductImages,
  ProductDetails,
  TitleAndPriceContainer,
  TitleAndDescription,
  PriceBox,
  PriceText,
  SellerInfo,
  SellerAvatar,
  ProductSpecsContainer,
  BottomTab,
  ShareInterestButton,
  ModalBackdrop,
  ModalContent,
  NoteTextArea,

  SpecGrid,
  SpecCategory,
  SpecItem,
};
