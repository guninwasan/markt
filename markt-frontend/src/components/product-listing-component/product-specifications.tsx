import {
  ProductSpecsContainer,
  SpecGrid,
  SpecCategory,
  SpecItem,
} from "./product-listing-component.styles";

const ProductSpecs = ({ specs }: any) => {
  return (
    <ProductSpecsContainer>
      <h2>Product Specifications</h2>
      <SpecGrid>
        <SpecCategory>
          <h3>Basic Information</h3>
          <SpecItem>Condition: {specs.basicInfo.condition}</SpecItem>
          <SpecItem>Brand: {specs.basicInfo.brand}</SpecItem>
          <SpecItem>Model: {specs.basicInfo.model}</SpecItem>
          <SpecItem>Year: {specs.basicInfo.yearOfManufacture}</SpecItem>
        </SpecCategory>
        <SpecCategory>
          <h3>Appearance</h3>
          <SpecItem>Color: {specs.appearance.color}</SpecItem>
          <SpecItem>Dimensions: {specs.appearance.dimensions}</SpecItem>
          <SpecItem>Weight: {specs.appearance.weight}</SpecItem>
          <SpecItem>Material: {specs.appearance.material}</SpecItem>
        </SpecCategory>
        <SpecCategory>
          <h3>Performance</h3>
          <SpecItem>Battery Life: {specs.performance.batteryLife}</SpecItem>
          <SpecItem>
            Storage Capacity: {specs.performance.storageCapacity}
          </SpecItem>
          <SpecItem>Features: {specs.performance.additionalFeatures}</SpecItem>
        </SpecCategory>
        <SpecCategory>
          <h3>Warranty</h3>
          <SpecItem>{specs.warranty}</SpecItem>
        </SpecCategory>
      </SpecGrid>
    </ProductSpecsContainer>
  );
};

export { ProductSpecs };
