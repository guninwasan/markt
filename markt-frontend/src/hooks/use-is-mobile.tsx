import { useEffect, useState } from "react";

const MOBILE_BREAKPOINT = 770;

const useIsMobile = () => {
  const [isMobile, setIsMobile] = useState<null | boolean>(null);

  useEffect(() => {
    const handleResize = () => {
      setIsMobile(window.innerWidth <= MOBILE_BREAKPOINT);
    };

    window.addEventListener("resize", handleResize);

    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, []);

  return { isMobile };
};

export { useIsMobile };
