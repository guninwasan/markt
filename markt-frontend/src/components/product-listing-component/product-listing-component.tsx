import { useIsMobile } from "../../hooks";
import { ImageGallery } from "../image-gallery";
import { useEffect, useState } from "react";
import {
  ProductListingContainer,
  ProductImages,
  ProductDetails,
  TitleAndPriceContainer,
  TitleAndDescription,
  PriceBox,
  PriceText,
  SellerInfo,
  SellerAvatar,
  ProductSpecsContainer,
  BottomTab,
} from "./product-listing-component.styles";

const ProductListingComponent = () => {
  const { isMobile } = useIsMobile();
  const [isBottomTabVisible, setBottomTabVisible] = useState(false);

  useEffect(() => {
    const handleScroll = () => {
      setBottomTabVisible(window.scrollY > 100);
    };

    window.addEventListener("scroll", handleScroll);
    return () => {
      window.removeEventListener("scroll", handleScroll);
    };
  }, []);

  return (
    <>
      <ProductListingContainer isMobile={isMobile}>
        <ProductImages>
          <ImageGallery
            images={[
              "https://via.placeholder.com/150",
              "https://via.placeholder.com/150",
              "https://via.placeholder.com/150",
              "https://via.placeholder.com/150",
              "https://via.placeholder.com/150",
              "https://via.placeholder.com/150",
            ]}
          />
        </ProductImages>
        <ProductDetails isMobile={isMobile}>
          <TitleAndPriceContainer isMobile={isMobile}>
            <TitleAndDescription>
              <h1>Product Title</h1>
              <p>
                Description description description description description
                description description description description description
                description description description description description
                description description description
              </p>
            </TitleAndDescription>
            <PriceBox>
              <PriceText>CAD $$$</PriceText>
              <div>Negotiable / Non-negotiable</div>
              <SellerInfo>
                <SellerAvatar
                  src="https://via.placeholder.com/50"
                  alt="Seller Avatar"
                />
                <div>
                  <div>NAME BLURRED</div>
                  <div>SELLER ID # 1234</div>
                  <a href="#">See all seller reviews</a>
                </div>
              </SellerInfo>
            </PriceBox>
          </TitleAndPriceContainer>
          <ProductSpecsContainer>
            <h2>Product Specifications</h2>
            <ul>
              <li>Specification 1</li>
              <li>Specification 2</li>
              <li>Specification 3</li>
              <li>Specification 4</li>
            </ul>
          </ProductSpecsContainer>
        </ProductDetails>
      </ProductListingContainer>
      {isMobile && (
        <BottomTab isVisible={isBottomTabVisible} test-id="bottom-tab">
          <SellerInfo>
            <SellerAvatar
              src="https://via.placeholder.com/50"
              alt="Seller Avatar"
            />
            <div>
              <div>NAME BLURRED</div>
              <div>SELLER ID # 1234</div>
              <a href="#">See all seller reviews</a>
            </div>
          </SellerInfo>
        </BottomTab>
      )}
    </>
  );
};

export { ProductListingComponent };
