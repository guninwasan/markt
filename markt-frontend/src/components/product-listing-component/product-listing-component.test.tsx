import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { ProductListingComponent } from "./product-listing-component";
import { useIsMobile } from "../../hooks";

jest.mock("../../hooks", () => ({
  useIsMobile: jest.fn(),
}));

describe("ProductListingComponent", () => {
  const mockUseIsMobile = useIsMobile as jest.Mock;

  beforeEach(() => {
    mockUseIsMobile.mockReturnValue({ isMobile: false });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly on desktop", () => {
    const { container } = render(<ProductListingComponent />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders correctly on mobile", () => {
    mockUseIsMobile.mockReturnValue({ isMobile: true });
    render(<ProductListingComponent />);
    expect(screen.getByText("Product Title")).toBeInTheDocument();
    expect(screen.getByText("CAD $$$")).toBeInTheDocument();
    expect(screen.getByText("Specification 1")).toBeInTheDocument();
  });

  it("renders product details correctly", () => {
    render(<ProductListingComponent />);
    expect(screen.getByText("Product Title")).toBeInTheDocument();
    expect(screen.getByText(/Description description/)).toBeInTheDocument();
    expect(screen.getByText("CAD $$$")).toBeInTheDocument();
    expect(screen.getByText("NAME BLURRED")).toBeInTheDocument();
    expect(screen.getByText("SELLER ID # 1234")).toBeInTheDocument();
    expect(screen.getByText("See all seller reviews")).toBeInTheDocument();
  });

  it("renders product specifications correctly", () => {
    render(<ProductListingComponent />);
    expect(screen.getByText("Product Specifications")).toBeInTheDocument();
    expect(screen.getByText("Specification 1")).toBeInTheDocument();
    expect(screen.getByText("Specification 2")).toBeInTheDocument();
    expect(screen.getByText("Specification 3")).toBeInTheDocument();
    expect(screen.getByText("Specification 4")).toBeInTheDocument();
  });
});
