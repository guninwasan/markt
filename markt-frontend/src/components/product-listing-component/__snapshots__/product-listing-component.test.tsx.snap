// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListingComponent renders correctly on desktop 1`] = `
<div
  class="sc-dntaoT itDxOn"
>
  <div
    class="sc-ivxoEo fdjjXu"
  >
    <div
      class="sc-blHHSb iAZfxc"
    >
      <div
        class="sc-gtLWhw Kyoay"
      >
        <img
          alt="Image 1"
          class="sc-fAUdSK eTRGma"
          src="https://via.placeholder.com/150"
        />
      </div>
      <div
        class="sc-gtLWhw Kyoay"
      >
        <img
          alt="Image 2"
          class="sc-fAUdSK eTRGma"
          src="https://via.placeholder.com/150"
        />
      </div>
      <div
        class="sc-gtLWhw Kyoay"
      >
        <img
          alt="Image 3"
          class="sc-fAUdSK eTRGma"
          src="https://via.placeholder.com/150"
        />
      </div>
      <div
        class="sc-gtLWhw Kyoay"
      >
        <img
          alt="Image 4"
          class="sc-fAUdSK eTRGma"
          src="https://via.placeholder.com/150"
        />
      </div>
      <div
        class="sc-gtLWhw Kyoay"
      >
        <img
          alt="Image 5"
          class="sc-fAUdSK eTRGma"
          src="https://via.placeholder.com/150"
        />
      </div>
      <div
        class="sc-gtLWhw Kyoay"
      >
        <img
          alt="Image 6"
          class="sc-fAUdSK eTRGma"
          src="https://via.placeholder.com/150"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-ghWlax dxtQXs"
  >
    <div
      class="sc-kLhKbu eocVKy"
    >
      <div
        class="sc-ixGGxD XXNEF"
      >
        <h1>
          Product Title
        </h1>
        <p>
          Description description description description description description description description description description description description description description description description description description
        </p>
      </div>
      <div
        class="sc-khLCKb ikiXMJ"
      >
        <div
          class="sc-dstKZu cqYejV"
        >
          CAD $$$
        </div>
        <div>
          Negotiable / Non-negotiable
        </div>
        <div
          class="sc-keTIit kYGxmp"
        >
          <img
            alt="Seller Avatar"
            class="sc-dpBQxM eOXdc"
            src="https://via.placeholder.com/50"
          />
          <div>
            <div>
              NAME BLURRED
            </div>
            <div>
              SELLER ID # 1234
            </div>
            <a
              href="#"
            >
              See all seller reviews
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-jwIPbr hmWefI"
    >
      <h2>
        Product Specifications
      </h2>
      <ul>
        <li>
          Specification 1
        </li>
        <li>
          Specification 2
        </li>
        <li>
          Specification 3
        </li>
        <li>
          Specification 4
        </li>
      </ul>
    </div>
  </div>
</div>
`;
